src/firefly/java/edu/caltech/ipac/firefly/server/query/ObsCorePackager.java

```diff
@@ -240,4 +288,137 @@ else if (obs_title != null) { //option 2, if productTitleTemplate is not availab
         if (file_name.length() == 0) file_name = "file"; //fallback option
         return file_name;
     }
+
+
+    //the functions and classes below parse the service descriptors for cutouts to create a cutout product URL
+
+    private static CutoutServiceDescriptor createCutoutSerDefUrl(DataGroup[] groups, String serviceDefId, String ra, String dec, String cutoutValue) {
+        for (DataGroup dg : groups) {
+            for (ResourceInfo ri : dg.getResourceInfos()) {
+                if ("adhoc:service".equals(ri.getUtype()) && serviceDefId.equals(ri.getID())) {
+                    CutoutServiceDescriptor result = parseCutoutServiceDescriptor(ri, ra, dec, cutoutValue);
+                    return result;
+                }
+            }
+        }
+        return null;
+    }
+
+
+    private static CutoutServiceDescriptor parseCutoutServiceDescriptor(ResourceInfo serviceDescriptor, String ra, String dec, String cutoutValue) {
+        String accessUrl = null;
+        String primaryParamName = null;
+        StringBuilder queryString = new StringBuilder();
+        List<MissingParam> missingParams = new ArrayList<>();
+
+        List<GroupInfo> groups = serviceDescriptor.getGroups();
+        List<ParamInfo> params = serviceDescriptor.getParams();
+
+        // Step 1: Extract accessURL
+        for (ParamInfo param : params) {
+            if ("accessURL".equals(param.getKeyName())) {
+                accessUrl = param.getStringValue();
+                break;
+            }
+        }
+        if (accessUrl == null || accessUrl.isEmpty()) {
+            return new CutoutServiceDescriptor("", new ArrayList<>()); // Invalid service descriptor
+        }
+
+        // Step 2: Locate the "inputParams" group
+        for (GroupInfo group : groups) {
+            if (!"inputParams".equals(group.getName())) continue;
+
+            List<ParamInfo> inputParams = group.getParamInfos();
+
+            //Check for xtype = "CIRCLE"
+            for (ParamInfo param : inputParams) {
+                if ("CIRCLE".equals(param.getXType())) {
```

Comment:
`equalsIgnoreCase`

You should use this everywhere for matching anything in the datalink fire.